**Discretización**

- Morosidad : moroso y no moroso (ML)
- fraude: fraude y no fraude (ML)
- células cancerígenas : presencia o no (deep learning) - redes neuronales convolucionales

**Naive bayes**: basado en la **probabilidad condicional**

In [ ]:
x         y                  y_estimado
25        1 (reclama)            0
35        0                      0
41        0                      0
28        1 (reclama)            0
29        1 (reclama)            0
.
.
.
10k casos

In [ ]:
Pr(x=¿?|y=1) = ¿?
Pr(x=25|y=1) = 0.02
Pr(x=26|y=1) = 0.03
Pr(x=27|y=1) = 0.01
Pr(x=28|y=1) = 0.04

**¿Cómo soluciono?**
- discretización

In [ ]:
1= 20 -30
2= 31 -45
3= +45

In [ ]:
Pr(x=1|y=1) = 0.5
Pr(x=2|y=1) = 0.3
Pr(x=3|y=1) = 0.2

**árboles de decisión**

**impureza** error del árbol (entropia y gini)

- discretización por entropia.

técnicas de discretización:

1. discretización por intervalos de igual amplitud
2. discretización por intervalos de igual frecuencia (cuantil)
3. discretización por k-means
4. discretización por entropía

**discretización por intervalos de igual amplitud**

In [2]:
from sklearn.preprocessing import KBinsDiscretizer
import math as m
import pandas as pd
import numpy as np

In [3]:
data = pd.read_csv("D:/Python/1. Nivel I/7/datos/data.csv")

In [4]:
data.head()

Unnamed: 0      rsrp  redireccion       cqi  incidentes
0           0 -116.5000  90890.42778  8.138370           0
1           1 -115.1875  91193.46144  9.140266           0
2           2 -119.1250  91877.50918  7.808917           0
3           3 -111.8125  92049.63360  8.878271           0
4           4 -115.1875  92275.58028  9.051622           0

In [5]:
data.drop(columns="Unnamed: 0",axis=1, inplace=True)

In [6]:
data.head()

rsrp  redireccion       cqi  incidentes
0 -116.5000  90890.42778  8.138370           0
1 -115.1875  91193.46144  9.140266           0
2 -119.1250  91877.50918  7.808917           0
3 -111.8125  92049.63360  8.878271           0
4 -115.1875  92275.58028  9.051622           0

In [7]:
data["cqi"].min()

4.0703296661377

In [8]:
data["cqi"].max()

11.62320369482

In [9]:
rango = data["cqi"].max()-data["cqi"].min(); print(rango)

7.552874028682301


**Cómo determino el número de categorías?**

In [14]:
sturges = round(1+m.log2(len(data)),0); print(sturges)

12.0


In [15]:
rango/sturges

0.6294061690568584

In [17]:
((rango/sturges)*2)+4.07

5.328812338113718

In [ ]:
1 = 4.07 - 4.6994
2 = 4.6995 - 5.3288
.
.
.
.
12 =         - 11.6232

**Nota**:este proceso de discretización estás afectado por la presencia de valores atípicos. No es recomendable aplicarlo cuando ustedes noten la presencia de valores atípicos.

In [19]:
amplitud = KBinsDiscretizer(n_bins=12, #número de categorías de la variable
                           encode="ordinal", # son de naturaleza ordinal
                           strategy="uniform") #uniform = discretizión por intervalos de igual amplitud

In [20]:
data["cqi_disc_amplitud"]=amplitud.fit_transform(data[["cqi"]])

In [22]:
data.head()

rsrp  redireccion       cqi  incidentes  cqi_disc_amplitud
0 -116.5000  90890.42778  8.138370           0                6.0
1 -115.1875  91193.46144  9.140266           0                8.0
2 -119.1250  91877.50918  7.808917           0                5.0
3 -111.8125  92049.63360  8.878271           0                7.0
4 -115.1875  92275.58028  9.051622           0                7.0

**discretización por intervalosd e igual frecuencia o discretización cuantil**

In [23]:
datan = [4,3,2,6,8,2,5,6]

**Paso 1**: ordenar de forma ascendente los valores de la variable

In [24]:
datan = [2,2,3,4,5,6,6,8]

**Paso 2** cuartilas [0.25,0.50, 0.75, 1]

In [ ]:
categoría 1 = [2,2]
categoría 2 = [3,4]
categoría 3 = [5,6,6]
categoría 4 = [8]

In [25]:
cuartil = KBinsDiscretizer(n_bins=4, #número de categorías de la variable
                           encode="ordinal", # son de naturaleza ordinal
                           strategy="quantile") 

In [26]:
data["cqi_cuartil"]=cuartil.fit_transform(data[["cqi"]])

In [27]:
data.head()

rsrp  redireccion       cqi  incidentes  cqi_disc_amplitud  cqi_cuartil
0 -116.5000  90890.42778  8.138370           0                6.0          1.0
1 -115.1875  91193.46144  9.140266           0                8.0          2.0
2 -119.1250  91877.50918  7.808917           0                5.0          0.0
3 -111.8125  92049.63360  8.878271           0                7.0          2.0
4 -115.1875  92275.58028  9.051622           0                7.0          2.0

**Discretización por KMeans**

<center><img src="https://miro.medium.com/max/963/1*Qu9vtRFFrQ1_4jj4Nu__BQ.jpeg" width="700"></center>

In [28]:
datos = [23,25,22,29,36,41,38,18]

In [29]:
c1 = 26
c2 = 32

In [30]:
data_cl = [1,1,1,2,2,2,2,1]

In [36]:
c1 = 22
c2 = 36

In [31]:
data_cl = [1,1,1,1,2,2,2,1] #hasta que se dejen de reasignar las observaciones a otros clústers

In [34]:
c1 = 23.4
c2 = 38.3

In [35]:
data_cl = [1,1,1,1,2,2,2,1] #hasta que se dejen de reasignar las observaciones a otros clústers

In [37]:
df = pd.DataFrame(datos, columns=["edad"])

In [38]:
df

edad
0    23
1    25
2    22
3    29
4    36
5    41
6    38
7    18

In [41]:
kmeans = KBinsDiscretizer(n_bins=2,
                           encode="ordinal",
                           strategy="kmeans")

In [42]:
df["edad_kmeans"]=kmeans.fit_transform(df[["edad"]])+1

In [43]:
df

edad  edad_kmeans
0    23          1.0
1    25          1.0
2    22          1.0
3    29          1.0
4    36          2.0
5    41          2.0
6    38          2.0
7    18          1.0

In [ ]:
data_cl = [1,1,1,1,2,2,2,1]

In [45]:
kmeans = KBinsDiscretizer(n_bins=2,
                           encode="ordinal",
                           strategy="kmeans")

In [49]:
data["cqi_kmeans"]=kmeans.fit_transform(data[["cqi"]])+1

In [50]:
data.head()

rsrp  redireccion       cqi  incidentes  cqi_disc_amplitud  \
0 -116.5000  90890.42778  8.138370           0                6.0   
1 -115.1875  91193.46144  9.140266           0                8.0   
2 -119.1250  91877.50918  7.808917           0                5.0   
3 -111.8125  92049.63360  8.878271           0                7.0   
4 -115.1875  92275.58028  9.051622           0                7.0   

   cqi_cuartil  cqi_kmeans  
0          1.0         1.0  
1          2.0         2.0  
2          0.0         1.0  
3          2.0         2.0  
4          2.0         2.0

In [ ]:
#############################################################################

**Ingeniero de datos**

**Canalización de preprocesamiento** (Pipeline)

- Conocer las técnicas de ingeniería de características y prepocesamiento de datos.

**Científico de datos**

**Canalización de modelamiento** (Pipeline)

- Conocer los modelos de aprendizaje (algoritmo, evaluación , optimización de hiperparámetros, interpretabilidad)

In [ ]:
#############################################################################

**Discretización por entropia o discretización basada en árboles**

- basada en un modelo de aprendizaje: árboles de decisión
- la discretización de la variable x se realizará en función a la variable y (target)

In [ ]:
x         y                  
25        1 (reclama)            
35        0                     
41        0                     
28        1 (reclama)            
29        1 (reclama)           
.
.
.
10k casos

In [51]:
r= 4000
nr= 6000
t=10000

In [55]:
impureza_ini = abs(((r/t)*m.log(r/t))+((nr/t)*m.log(nr/t))); print(impureza_ini)

0.6730116670092565


**Se va a discretizar la variable x utilizando diferentes puntos de corte (puntos de discretización) y ganará el que reduzca la impureza del árbol**

- La impureza inicial del árbol es 0.673 y la impureza final usando el punto de corte "30 años" es 0.672
- La impureza inicial del árbol es 0.673 y la impureza final usando el punto de corte "38 años" es 0.35
- La impureza inicial del árbol es 0.673 y la impureza final usando el punto de corte "40 años" es 0.27

**Seleccionamos el punto de corte de discretización** que reduce la impureza inicial de forma más significativa

In [56]:
import os
os.getcwd()

'C:\\Users\\SOCIAL DATA'

In [57]:
from MDLP import MDLP_Discretizer

In [58]:
data.head()

rsrp  redireccion       cqi  incidentes  cqi_disc_amplitud  \
0 -116.5000  90890.42778  8.138370           0                6.0   
1 -115.1875  91193.46144  9.140266           0                8.0   
2 -119.1250  91877.50918  7.808917           0                5.0   
3 -111.8125  92049.63360  8.878271           0                7.0   
4 -115.1875  92275.58028  9.051622           0                7.0   

   cqi_cuartil  cqi_kmeans  
0          1.0         1.0  
1          2.0         2.0  
2          0.0         1.0  
3          2.0         2.0  
4          2.0         2.0

In [59]:
x=data.iloc[:,0:3].values

In [60]:
y=data.iloc[:,3].values

In [61]:
ba=[2]#posicion de la columna a discretizar

In [62]:
n_predictores = np.array(ba)

In [63]:
n_predictores

array([2])

In [64]:
discretizer = MDLP_Discretizer(features=n_predictores)

In [65]:
discretizer.fit(x,y)

In [69]:
data["cqi_entropia"]=discretizer.transform(x)[:,2]+1

In [70]:
data.head()

rsrp  redireccion       cqi  incidentes  cqi_disc_amplitud  \
0 -116.5000  90890.42778  8.138370           0                6.0   
1 -115.1875  91193.46144  9.140266           0                8.0   
2 -119.1250  91877.50918  7.808917           0                5.0   
3 -111.8125  92049.63360  8.878271           0                7.0   
4 -115.1875  92275.58028  9.051622           0                7.0   

   cqi_cuartil  cqi_kmeans  cqi_entropia  
0          1.0         1.0           1.0  
1          2.0         2.0           2.0  
2          0.0         1.0           1.0  
3          2.0         2.0           2.0  
4          2.0         2.0           2.0

In [ ]:
#################################################################################

**Escalimiento de variables**

- algoritmos o modelos de agrupamiento de observaciones como K-means & AGNES (clúster jerárquico aglomerativo)
- técnicas de reducción de dimensionalidad: PCA & FA.

Imputo (valores faltantes) - Escalo (escala de medida) - Reducces (multicolinealidad de datos) 

**dos técnicas**

**Normalización**

In [71]:
edad = [23,25,27,39,25,30,28,32]

In [72]:
mini=23
maxi=39

In [76]:
(39-mini)/(maxi-mini)

1.0

**Nota**: transforma el espacio original de comportamiento de los datos

**Estandarización**

In [83]:
import statistics
st_dev=statistics.stdev(edad)

In [84]:
st_mean=statistics.mean(edad)

In [87]:
(27-st_mean)/st_dev

-0.31879821581291556

**Nota**: sin afectar el comportamiento (transformar) original de la variable

In [113]:
import pandas as pd
from sklearn.model_selection import train_test_split

In [114]:
df = pd.read_csv("D:/Python/1. Nivel I/9/datos/data_credit.csv",
                sep=",",
                encoding="ISO-8859-1")

In [115]:
df.head()

duration  chist  reason    camt  telephne  lenemp  instrate  perstat  \
0       6.0    5.0     4.0  1169.0       0.0     4.0       4.0      3.0   
1      48.0    3.0     4.0  5951.0       1.0     2.0       2.0      2.0   
2      12.0    5.0     7.0  2096.0       1.0     3.0       2.0      3.0   
3      42.0    3.0     3.0  7882.0       1.0     3.0       2.0      3.0   
4      24.0    4.0     1.0  4870.0       1.0     2.0       3.0      3.0   

   residlen  prpownr   age  housng  numcred  emptype  numliab  customer  
0       4.0      1.0  67.0     2.0      2.0      3.0      1.0       1.0  
1       2.0      1.0  22.0     2.0      1.0      3.0      1.0       2.0  
2       3.0      1.0  49.0     2.0      1.0      2.0      2.0       1.0  
3       4.0      2.0  45.0     3.0      1.0      3.0      2.0       1.0  
4       4.0      4.0  53.0     3.0      2.0      3.0      2.0       2.0

In [116]:
df.tail()

duration  chist  reason    camt  telephne  lenemp  instrate  perstat  \
995      12.0    3.0     3.0  1736.0       1.0     3.0       3.0      2.0   
996      30.0    3.0     2.0  3857.0       0.0     2.0       4.0      1.0   
997      12.0    3.0     4.0   804.0       1.0     4.0       4.0      3.0   
998      45.0    3.0     4.0  1845.0       0.0     2.0       4.0      3.0   
999      45.0    5.0     2.0  4576.0       1.0     5.0       3.0      3.0   

     residlen  prpownr   age  housng  numcred  emptype  numliab  customer  
995       4.0      1.0  31.0     2.0      1.0      2.0      1.0       1.0  
996       4.0      2.0  40.0     2.0      1.0      4.0      1.0       1.0  
997       4.0      3.0  38.0     2.0      1.0      3.0      1.0       1.0  
998       4.0      4.0  23.0     3.0      1.0      3.0      1.0       2.0  
999       4.0      3.0  27.0     2.0      1.0      3.0      1.0       1.0

In [92]:
df = df[["duration","camt","age","customer"]]

In [94]:
df.head()

duration    camt   age  customer
0       6.0  1169.0  67.0       1.0
1      48.0  5951.0  22.0       2.0
2      12.0  2096.0  49.0       1.0
3      42.0  7882.0  45.0       1.0
4      24.0  4870.0  53.0       2.0

In [96]:
x=df.iloc[:,0:3].values
y=df.iloc[:,3].values

In [97]:
x_train, x_test, y_train, y_test= train_test_split(x,
                                                  y,
                                                  test_size=0.3,
                                                  random_state=2023)

In [98]:
from sklearn.preprocessing import MinMaxScaler

In [99]:
minmax=MinMaxScaler()

In [100]:
x_train_norm= minmax.fit_transform(x_train)

In [101]:
x_train_norm

array([[0.02941176, 0.02321998, 0.63636364],
       [0.11764706, 0.04033234, 0.        ],
       [0.11764706, 0.10360955, 0.6       ],
       ...,
       [0.29411765, 0.25129306, 0.21818182],
       [0.20588235, 0.18498955, 0.32727273],
       [0.29411765, 0.06734896, 0.25454545]])

In [102]:
x_test_norm=minmax.transform(x_test)

In [103]:
x_test_norm

array([[0.29411765, 0.05496864, 0.16363636],
       [0.11764706, 0.18377902, 0.18181818],
       [0.02941176, 0.01089468, 0.07272727],
       [0.51470588, 0.12759987, 0.09090909],
       [0.29411765, 0.12176736, 0.14545455],
       [0.16176471, 0.13634863, 0.29090909],
       [0.11764706, 0.10030813, 0.16363636],
       [0.64705882, 0.58935842, 0.14545455],
       [0.29411765, 0.06707384, 0.4       ],
       [0.35294118, 0.13717398, 0.18181818],
       [0.47058824, 0.39517993, 0.63636364],
       [0.11764706, 0.03499505, 0.03636364],
       [0.29411765, 0.21563772, 0.16363636],
       [0.38235294, 0.15225047, 0.50909091],
       [0.29411765, 0.12286783, 0.05454545],
       [0.38235294, 0.15081985, 0.10909091],
       [0.38235294, 0.22003962, 0.16363636],
       [0.29411765, 0.07202597, 0.56363636],
       [0.20588235, 0.3995268 , 0.58181818],
       [0.29411765, 0.14636294, 0.27272727],
       [0.29411765, 0.07444701, 0.65454545],
       [0.33823529, 0.46016287, 0.14545455],
       [0.

In [104]:
from sklearn.preprocessing import StandardScaler

In [105]:
std=StandardScaler()

In [106]:
x_train_std = std.fit_transform(x_train)

In [107]:
x_train_std

array([[-1.21234165, -0.88830937,  1.61422557],
       [-0.71865977, -0.77970153, -1.45965337],
       [-0.71865977, -0.37809697,  1.43857535],
       ...,
       [ 0.268704  ,  0.55921314, -0.40575202],
       [-0.22497789,  0.13840141,  0.12119866],
       [ 0.268704  , -0.60823384, -0.2301018 ]])

In [108]:
x_test_std=std.transform(x_test)

In [109]:
x_test_std

array([[ 2.68703996e-01, -6.86808649e-01, -6.69227360e-01],
       [-7.18659769e-01,  1.30718542e-01, -5.81402247e-01],
       [-1.21234165e+00, -9.66534953e-01, -1.10835292e+00],
       [ 1.50290870e+00, -2.25836461e-01, -1.02052781e+00],
       [ 2.68703996e-01, -2.62853925e-01, -7.57052472e-01],
       [-4.71818828e-01, -1.70310266e-01, -5.44515699e-02],
       [-7.18659769e-01, -3.99050253e-01, -6.69227360e-01],
       [ 2.24343153e+00,  2.70482914e+00, -7.57052472e-01],
       [ 2.68703996e-01, -6.09979951e-01,  4.72499107e-01],
       [ 5.97825251e-01, -1.65071945e-01, -5.81402247e-01],
       [ 1.25606776e+00,  1.47242699e+00,  1.61422557e+00],
       [-7.18659769e-01, -8.13576000e-01, -1.28400315e+00],
       [ 2.68703996e-01,  3.32917706e-01, -6.69227360e-01],
       [ 7.62385878e-01, -6.93852942e-02,  9.99449784e-01],
       [ 2.68703996e-01, -2.55869497e-01, -1.19617804e+00],
       [ 7.62385878e-01, -7.84650494e-02, -9.32702698e-01],
       [ 7.62385878e-01,  3.60855414e-01

In [110]:
df_corr = pd.DataFrame(x_train,
                      columns=df.columns.tolist()[0:3])

In [112]:
df_corr.corr(method="pearson")

duration      camt       age
duration  1.000000  0.639104 -0.052212
camt      0.639104  1.000000  0.051650
age      -0.052212  0.051650  1.000000